File: library/src/test/java/com/wultra/android/sslpinning/CertStoreUpdateTest.java
/*
 * Copyright 2018 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 */

package com.wultra.android.sslpinning;

import android.support.annotation.NonNull;

import com.wultra.android.sslpinning.integration.DefaultUpdateObserver;
import com.wultra.android.sslpinning.interfaces.ECPublicKey;
import com.wultra.android.sslpinning.interfaces.SignedData;
import com.wultra.android.sslpinning.service.RemoteDataProvider;
import com.wultra.android.sslpinning.service.RemoteDataRequest;
import com.wultra.android.sslpinning.service.RemoteDataResponse;

import org.jetbrains.annotations.NotNull;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.modules.junit4.PowerMockRunner;

import java.net.URL;
import java.util.Base64;
import java.util.Date;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import static java.util.Collections.emptyMap;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * Unit tests for {@link CertStore} updates.
 *
 * @author Tomas Kypta, tomas.kypta@wultra.com
 */
@RunWith(PowerMockRunner.class)
public class CertStoreUpdateTest extends CommonJavaTest {

    @Test
    public void testCorrectUpdate() throws Exception {
        when(cryptoProvider.ecdsaValidateSignatures(any(SignedData.class), any(ECPublicKey.class)))
                .thenAnswer(invocation -> true);

        String pinningJsonUrl = "https://gist.githubusercontent.com/hvge/7c5a3f9ac50332a52aa974d90ea2408c/raw/34866234bbaa3350dc0ddc5680a65a6f4e7c549e/ssl-pinning-signatures.json";
        UpdateResult updateResult = performForcedUpdate(pinningJsonUrl);
        assertEquals(UpdateResult.OK, updateResult);
    }

    @Test
    public void testInvalidSignatureUpdate() throws Exception {
        when(cryptoProvider.ecdsaValidateSignatures(any(SignedData.class), any(ECPublicKey.class)))
                .thenAnswer(invocation -> false);

        String pinningJsonUrl = "https://gist.githubusercontent.com/hvge/7c5a3f9ac50332a52aa974d90ea2408c/raw/34866234bbaa3350dc0ddc5680a65a6f4e7c549e/ssl-pinning-signatures.json";
        UpdateResult updateResult = performForcedUpdate(pinningJsonUrl);
        assertEquals(UpdateResult.INVALID_SIGNATURE, updateResult);
    }

    @Test
    public void testExpiredUpdate() throws Exception {
        when(cryptoProvider.ecdsaValidateSignatures(any(SignedData.class), any(ECPublicKey.class)))
                .thenAnswer(invocation -> false);

        String pinningJsonUrl = "https://gist.githubusercontent.com/TomasKypta/5a6d99fe441a8c0d201b673d88e223a6/raw/0d12746cad1247ebf9a5b1706afabf8486a7a62e/ssl-pinning-signatures_expired.json";
        UpdateResult updateResult = performForcedUpdate(pinningJsonUrl);
        assertEquals(UpdateResult.STORE_IS_EMPTY, updateResult);
    }

    @Test
    public void testUpdateSignatureGithub() throws Exception {
        String publicKey = "BC3kV9OIDnMuVoCdDR9nEA/JidJLTTDLuSA2TSZsGgODSshfbZg31MS90WC/HdbU/A5WL5GmyDkE/iks6INv+XE=";
        byte[] publicKeyBytes = java.util.Base64.getDecoder().decode(publicKey);

        CertStoreConfiguration config = TestUtils.getCertStoreConfiguration(
                new Date(),
                new String[]{"github.com"},
                new URL("https://gist.githubusercontent.com/hvge/7c5a3f9ac50332a52aa974d90ea2408c/raw/34866234bbaa3350dc0ddc5680a65a6f4e7c549e/ssl-pinning-signatures.json"),
                publicKeyBytes,
                null);
        RemoteDataProvider remoteDataProvider = mock(RemoteDataProvider.class);
        String jsonData =
                "{\n" +
                        "  \"fingerprints\": [\n" +
                        "    {\n" +
                        "      \"name\" : \"github.com\",\n" +
                        "      \"fingerprint\" : \"trmmrz6GbL4OajB+fdoXOzcrLTrD8GrxX5dxh3OEgAg=\",\n" +
                        "      \"expires\" : 1652184000,\n" +
                        "      \"signature\" : \"MEUCIQCs1y/nyrKh4+2DIuX/PufUYiaVUdt2FBZQg6rBeZ/r4QIgNlT4owBwJ1ThrDsE0SwGipTNI74vP1vNyLNEwuXY4lE=\"\n" +
                        "    }\n" +
                        "  ]\n" +
                        "}";
        when(remoteDataProvider.getFingerprints(new RemoteDataRequest(emptyMap()))).thenReturn(new RemoteDataResponse(200, emptyMap(), jsonData.getBytes()));

        CertStore store = new CertStore(config, cryptoProvider, secureDataStore, remoteDataProvider);
        TestUtils.assignHandler(store, handler);
        TestUtils.updateAndCheck(store, UpdateMode.FORCED, UpdateResult.OK);
    }

    @Test
    public void testUpdateWithNoUpdateObserver() throws Exception {
        String publicKey = "BC3kV9OIDnMuVoCdDR9nEA/JidJLTTDLuSA2TSZsGgODSshfbZg31MS90WC/HdbU/A5WL5GmyDkE/iks6INv+XE=";
        byte[] publicKeyBytes = java.util.Base64.getDecoder().decode(publicKey);

        CertStoreConfiguration config = TestUtils.getCertStoreConfiguration(
                new Date(),
                new String[]{"github.com"},
                new URL("https://gist.githubusercontent.com/hvge/7c5a3f9ac50332a52aa974d90ea2408c/raw/c5b021db0fcd40b1262ab513bf375e4641834925/ssl-pinning-signatures.json"),
                publicKeyBytes,
                null);
        RemoteDataProvider remoteDataProvider = mock(RemoteDataProvider.class);
        String jsonData =
                "{\n" +
                        "  \"fingerprints\": [\n" +
                        "    {\n" +
                        "      \"name\" : \"github.com\",\n" +
                        "      \"fingerprint\" : \"trmmrz6GbL4OajB+fdoXOzcrLTrD8GrxX5dxh3OEgAg=\",\n" +
                        "      \"expires\" : 1652184000,\n" +
                        "      \"signature\" : \"MEUCIQCs1y/nyrKh4+2DIuX/PufUYiaVUdt2FBZQg6rBeZ/r4QIgNlT4owBwJ1ThrDsE0SwGipTNI74vP1vNyLNEwuXY4lE=\"\n" +
                        "    }\n" +
                        "  ]\n" +
                        "}";

        CountDownLatch latch = new CountDownLatch(1);
        when(remoteDataProvider.getFingerprints(new RemoteDataRequest(emptyMap()))).thenAnswer(
                invocation -> {
                    byte[] bytes = jsonData.getBytes();
                    latch.countDown();
                    return new RemoteDataResponse(200, emptyMap(), bytes);
                });

        CertStore store = new CertStore(config, cryptoProvider, secureDataStore, remoteDataProvider);
        TestUtils.assignHandler(store, handler);


        store.update(UpdateMode.FORCED, new DefaultUpdateObserver() {
            @Override
            public void onUpdateStarted(@NotNull UpdateType type) {
                assertEquals(UpdateType.DIRECT, type);
                super.onUpdateStarted(type);
            }

            @Override
            public void onUpdateFinished(@NotNull UpdateType type, @NotNull UpdateResult result) {
                assertEquals(UpdateType.DIRECT, type);
                super.onUpdateFinished(type, result);
            }

            @Override
            public void handleFailedUpdate(@NotNull UpdateType type, @NotNull UpdateResult result) {
                fail();
            }

            @Override
            public void continueExecution() {

            }
        });
        assertTrue(latch.await(2, TimeUnit.SECONDS));
    }

    @NonNull
    private UpdateResult performForcedUpdate(String pinningJsonUrl) throws Exception {
        String publicKey = "BC3kV9OIDnMuVoCdDR9nEA/JidJLTTDLuSA2TSZsGgODSshfbZg31MS90WC/HdbU/A5WL5GmyDkE/iks6INv+XE=";
        byte[] publicKeyBytes = Base64.getDecoder().decode(publicKey);

        CertStoreConfiguration config = TestUtils.getCertStoreConfiguration(
                new Date(),
                new String[]{"github.com"},
                new URL(pinningJsonUrl),
                publicKeyBytes,
                null);
        CertStore store = new CertStore(config, cryptoProvider, secureDataStore);
        TestUtils.assignHandler(store, handler);

        return TestUtils.updateAndCheck(store, UpdateMode.FORCED, null);
    }
}

File: library/src/test/java/com/wultra/android/sslpinning/CertStoreUpdateTest.kt
/*
 * Copyright 2018 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 */
package com.wultra.android.sslpinning

import com.wultra.android.sslpinning.integration.DefaultUpdateObserver
import com.wultra.android.sslpinning.service.RemoteDataProvider
import com.wultra.android.sslpinning.service.RemoteDataResponse
import io.mockk.every
import io.mockk.mockk
import org.junit.Assert
import org.junit.Test
import java.net.URL
import java.util.Base64
import java.util.Date
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

/**
 * Unit tests for [CertStore] updates.
 *
 * @author Tomas Kypta, tomas.kypta@wultra.com
 */
class CertStoreUpdateTest : CommonKotlinTest() {

    @Test
    @Throws(Exception::class)
    fun testCorrectUpdate() {

        val remoteDataProvider: RemoteDataProvider = mockk()
        every { remoteDataProvider.getFingerprints(any()) } answers {
            RemoteDataResponse(
                200,
                mapOf(CertStore.RESPONSE_SIGNATURE_HEADER to "AAAA"),
                """
                    {
                      "fingerprints": [
                        {
                          "name" : "github.com",
                          "fingerprint" : "kqN/vV4hpTqVxxbhFE9EL1grlND6/Gc+tnF6TrUaiKc=",
                          "expires" : 2212460799,
                          "signature" : "MEUCICB69UpMPOdtrsR6XcJqHEh2L2RO4oSJ3SZ7BYnTBJbGAiEAnZ7rEWdMVGwa59Wx5QbAorEFxXH89Iu0CnqWa96Eda0="
                        }
                      ]
                    }
                """.toByteArray()
            )
        }

        every { cryptoProvider.ecdsaValidateSignature(any(), any()) } returns true

        val updateResult = performForcedUpdate(remoteDataProvider)
        Assert.assertEquals(UpdateResult.OK, updateResult)
    }

    @Test
    @Throws(Exception::class)
    fun testInvalidSignatureUpdate() {

        val remoteDataProvider: RemoteDataProvider = mockk()
        every { remoteDataProvider.getFingerprints(any()) } answers {
            RemoteDataResponse(
                200,
                mapOf(CertStore.RESPONSE_SIGNATURE_HEADER to "AAAA"),
                """
                    { 
                      "fingerprints": [{
                        "name" : "github.com",
                        "fingerprint" : "kqN/vV4hpTqVxxbhFE9EL1grlND6/Gc+tnF6TrUaiKc=",
                        "expires" : 2212460799,
                        "signature" : "MEQCIAiq/O5IbZ8K2SsZtDbpsvHWecxu4eLOSS7oOXjDk2KeAiBKkI53ESVByK+wKwaLA5LsEu8oonUHiYVM2zQtWf46DA=="
                      }]
                    }
                """.toByteArray()
            )
        }

        every { cryptoProvider.ecdsaValidateSignature(any(), any()) } returns false
        val updateResult = performForcedUpdate(remoteDataProvider)
        Assert.assertEquals(UpdateResult.INVALID_SIGNATURE, updateResult)
    }

    @Test
    @Throws(Exception::class)
    fun testExpiredUpdate() {

        val remoteDataProvider: RemoteDataProvider = mockk()
        every { remoteDataProvider.getFingerprints(any()) } answers {
            RemoteDataResponse(
                200,
                mapOf(CertStore.RESPONSE_SIGNATURE_HEADER to "AAAA"),
                """
                    {
                      "fingerprints": [
                        {
                          "name" : "github.com",
                          "fingerprint" : "MRFQDEpmASza4zPsP8ocnd5FyVREDn7kE3Fr/zZjwHQ=",
                          "expires" : 1531185600,
                          "signature" : "MEUCIQD8nGyux9GM8u3XCrRiuJj/N2eEuB0oiHzTEpGyy2gE9gIgYIRfyed6ykDzZbK1ougq1SoRW8UBe5q3VmWihHuL2JY="
                        }
                      ]
                    }
                """.toByteArray()
            )
        }

        every { cryptoProvider.ecdsaValidateSignature(any(), any()) } returns true

        val updateResult = performForcedUpdate(remoteDataProvider)
        Assert.assertEquals(UpdateResult.OK, updateResult)
    }

    @Test
    @Throws(Exception::class)
    fun testUpdateWithNoUpdateObserver() {
        val remoteDataProvider: RemoteDataProvider = mockk()
        val latch = CountDownLatch(1)
        every { remoteDataProvider.getFingerprints(any()) } answers {
            val bytes = """{
              "fingerprints": [
                {
                  "name" : "github.com",
                  "fingerprint" : "trmmrz6GbL4OajB+fdoXOzcrLTrD8GrxX5dxh3OEgAg=",
                  "expires" : 1652184000,
                  "signature" : "MEUCIQCs1y/nyrKh4+2DIuX/PufUYiaVUdt2FBZQg6rBeZ/r4QIgNlT4owBwJ1ThrDsE0SwGipTNI74vP1vNyLNEwuXY4lE="
                }
              ]
            }""".toByteArray()
            latch.countDown()
            RemoteDataResponse(200, mapOf(CertStore.RESPONSE_SIGNATURE_HEADER to "AAAA"), bytes)
        }
        val store = getCertStore(remoteDataProvider)
        TestUtils.assignHandler(store, handler)
        store.update(UpdateMode.FORCED, object : DefaultUpdateObserver() {
            override fun onUpdateStarted(type: UpdateType) {
                Assert.assertEquals(UpdateType.DIRECT, type)
                super.onUpdateStarted(type)
            }

            override fun onUpdateFinished(type: UpdateType, result: UpdateResult) {
                Assert.assertEquals(UpdateType.DIRECT, type)
                super.onUpdateFinished(type, result)
            }

            override fun handleFailedUpdate(type: UpdateType, result: UpdateResult) {
                Assert.fail()
            }

            override fun continueExecution() {}
        })
        Assert.assertTrue(latch.await(2, TimeUnit.SECONDS))
    }

    @Test
    @Throws(Exception::class)
    fun testMissingSignatureHeader() {
        // Arrange: valid response body, but no x-cert-pinning-signature header at all.
        val remoteDataProvider: RemoteDataProvider = mockk()
        every { remoteDataProvider.getFingerprints(any()) } answers {
            RemoteDataResponse(
                200,
                emptyMap(),
                """
                    {
                      "fingerprints": [
                        {
                          "name" : "github.com",
                          "fingerprint" : "kqN/vV4hpTqVxxbhFE9EL1grlND6/Gc+tnF6TrUaiKc=",
                          "expires" : 2212460799,
                          "signature" : "MEUCICB69UpMPOdtrsR6XcJqHEh2L2RO4oSJ3SZ7BYnTBJbGAiEAnZ7rEWdMVGwa59Wx5QbAorEFxXH89Iu0CnqWa96Eda0="
                        }
                      ]
                    }
                """.toByteArray()
            )
        }

        // Act
        val updateResult = performForcedUpdate(remoteDataProvider)

        // Assert: absent header must be treated as an invalid signature — security-critical.
        Assert.assertEquals(UpdateResult.INVALID_SIGNATURE, updateResult)
    }

    @Throws(Exception::class)
    private fun performForcedUpdate(remoteDataProvider: RemoteDataProvider): UpdateResult {
        val store = getCertStore(remoteDataProvider)
        TestUtils.assignHandler(store, handler)
        return TestUtils.updateAndCheck(store, UpdateMode.FORCED, null)
    }

    private fun getCertStore(remoteDataProvider: RemoteDataProvider): CertStore {
        val publicKeyBytes = Base64.getDecoder().decode("BC3kV9OIDnMuVoCdDR9nEA/JidJLTTDLuSA2TSZsGgODSshfbZg31MS90WC/HdbU/A5WL5GmyDkE/iks6INv+XE=")
        val config = TestUtils.getCertStoreConfiguration(
            Date(),
            arrayOf("github.com"),
            URL("https://test"),
            publicKeyBytes,
            null
        )
        return CertStore(config, cryptoProvider, secureDataStore, remoteDataProvider)
    }
}